ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -323,18 +323,36 @@ func (r *RayServiceReconciler) cleanUpServeConfigCache(rayServiceInstance *rayv1
 
 // shouldPrepareNewRayCluster checks if we need to generate a new pending cluster.
 func (r *RayServiceReconciler) shouldPrepareNewRayCluster(rayServiceInstance *rayv1alpha1.RayService, activeRayCluster *rayv1alpha1.RayCluster) bool {
-	shouldPrepareRayCluster := false
-
+	// Prepare new RayCluster if:
+	// 1. No active cluster and no pending cluster
+	// 2. No pending cluster, and the active RayCluster has changed.
 	if rayServiceInstance.Status.PendingServiceStatus.RayClusterName == "" {
-		// Prepare new RayCluster if:
-		// 1. No active and pending cluster
-		// 2. No pending cluster, and the active RayCluster has changed.
-		if activeRayCluster == nil || !utils.CompareJsonStruct(activeRayCluster.Spec, rayServiceInstance.Spec.RayClusterSpec) {
-			shouldPrepareRayCluster = true
+		if activeRayCluster == nil {
+			r.Log.Info("No active Ray cluster. RayService operator should prepare a new Ray cluster.")
+			return true
+		}
+		activeClusterHash := activeRayCluster.ObjectMeta.Annotations[common.RayServiceClusterHashKey]
+		goalClusterHash, err := utils.GenerateJsonHash(rayServiceInstance.Spec.RayClusterSpec)
+		if err != nil {
+			errContext := "Failed to serialize new RayCluster config. " +
+				"Manual config updates will NOT be tracked accurately. " +
+				"Please manually tear down the cluster and apply a new config."
+			r.Log.Error(err, errContext)
+		}
```

Comment:
do we return true or false given this error?

Reply:
I updated it to return `true` there. I think it's less surprising to crash loop early, rather than failing to update later.